# Data Cleaning Notebook
### Provides cleand data ready for Analysis & preprocessing steps

Real-world web-scraped data is inherently messy and needs a solid transformation pipeline before it can be fed into Analysis or Visualization tools

In [97]:
import pandas as pd
import numpy as np

In [98]:
data = pd.read_csv("Data/Raw/By_Filter/Scrapped_Data.csv", encoding='utf-8-sig')
data

,search_query,job_title,company_name,location,posting_date,job_url,years_exp_min,years_exp_max,career_level,education_level,salary,employment_type,job_categories,skills,job_description,job_requirements,scraped_at
0,IT-Software-Development,Call Center Agent /B1+ English Speakers/ e&UAE...,Etisalat Egypt,"Cairo, EG",posted 2 hours ago,https://wuzzuf.net/jobs/p/rxnmjewk9jch-call-ce...,0.0,4.0,Entry Level (Junior Level / Fresh Grad),Bachelor's Degree,12500 to 15000 EGP Per Month,Full Time | On-site,Customer Service/Support | IT/Software Develop...,Customer Service | Customer Support | Customer...,e& CX Solutions is hiring very good English up...,NaN,2026-04-19T02:03:20
1,IT-Software-Development,Mechatronics Engineer,Elhelb Smart Solutions,"Damietta, EG",posted 9 hours ago,https://wuzzuf.net/jobs/p/vb4tienz9ok2-mechatr...,0.0,2.0,Entry Level (Junior Level / Fresh Grad),Not Specified,"Confidential, Company Are Welcome To Give Huge...",Full Time | Part Time | Volunteering | On-site,Installation/Maintenance/Repair | IT/Software ...,Mechatronics | Accounting | Sales | Mechanical...,"- Install smart switches, dimmers, touch panel...",- Mechatronics Engineering is preferred — Elec...,2026-04-19T02:03:28
2,IT-Software-Development,Information Systems Specialist,Abou Seif Group,"Cairo, EG",posted 11 hours ago,https://wuzzuf.net/jobs/p/yijxsr0ymssu-informa...,3.0,5.0,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,Accounting/Finance | IT/Software Development,ERP | Accounting | Financial Analysis | Oracle...,Job Description\n\nWe are seeking a detail-ori...,Job Requirements\nBachelor’s degree in Account...,2026-04-19T02:03:35
3,IT-Software-Development,Product Designer,Rescounts,"Cairo, EG",posted 2 days ago,https://wuzzuf.net/jobs/p/yvumfgn0o4yp-product...,5.0,7.0,Experienced (Non-Manager),Not Specified,65000 to 75000 EGP Per Month,Full Time | Remote,Creative/Design/Art | IT/Software Development,Design | Product Design,"As a product designer, you will discover the b...",7+ years of professional experience with a por...,2026-04-19T02:03:41
4,IT-Software-Development,Senior Data Engineer,GB Corp,"Giza, EG",posted 2 days ago,https://wuzzuf.net/jobs/p/desxrlp7krfz-senior-...,3.0,7.0,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,IT/Software Development | Engineering - Teleco...,Computer Science | Information Technology (IT)...,"Responsible for designing, building, and maint...",Educational Requirements: Bachelor’s or Master...,2026-04-19T02:03:48
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
452,IT-Software-Development,Low Current Engineer,Roya Developments,"Alexandria, EG",posted 2 months ago,https://wuzzuf.net/jobs/p/o9jqu8sczz1v-low-cur...,2.0,2.0,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,IT/Software Development | Engineering - Teleco...,Low Current Engineer | ELV Systems | Fiber Net...,Follow-up of and monitor Fiber Networks and UT...,"Bachelor’s degree in computer science, Informa...",2026-04-19T02:57:22
453,IT-Software-Development,"UAT & Certification Specialist ""Outsource""",Efada Technology,"Cairo, EG",posted 2 months ago,https://wuzzuf.net/jobs/p/cy6ldbbvov7m-uat-cer...,2.0,3.0,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,Accounting/Finance | IT/Software Development |...,UAT | Testing | Automation | WordPress CMS | v...,Outsource at an Enterprise with Cards processi...,Experience / Skills Required\n\nBachelor’s deg...,2026-04-19T02:57:29
454,IT-Software-Development,Consultant - Command & Control Center,Confidential,"Cairo, EG",posted 2 months ago,https://wuzzuf.net/jobs/p/d2hhvrjfy3io-consult...,5.0,5.0,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,IT/Software Development,Professional Content | Reports and Insights | ...,Consultant – Command and Control Centre will s...,5+ years of consulting experience in Command &...,2026-04-19T02:57:34
455,IT-Software-Dev

In [99]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   search_query      457 non-null    object 
 1   job_title         457 non-null    object 
 2   company_name      457 non-null    object 
 3   location          457 non-null    object 
 4   posting_date      457 non-null    object 
 5   job_url           457 non-null    object 
 6   years_exp_min     456 non-null    float64
 7   years_exp_max     456 non-null    float64
 8   career_level      457 non-null    object 
 9   education_level   457 non-null    object 
 10  salary            457 non-null    object 
 11  employment_type   457 non-null    object 
 12  job_categories    457 non-null    object 
 13  skills            457 non-null    object 
 14  job_description   457 non-null    object 
 15  job_requirements  453 non-null    object 
 16  scraped_at        457 non-null    object 
dt

In [100]:
data.describe(include='all')

,search_query,job_title,company_name,location,posting_date,job_url,years_exp_min,years_exp_max,career_level,education_level,salary,employment_type,job_categories,skills,job_description,job_requirements,scraped_at
count,457,457,457,457,457,457,456.000000,456.000000,457,457,457,457,457,457,457,453,457
unique,1,392,234,10,30,457,NaN,NaN,4,2,79,18,62,453,452,450,457
top,IT-Software-Development,IT Specialist,Confidential,"Cairo, EG",posted 2 months ago,https://wuzzuf.net/jobs/p/rxnmjewk9jch-call-ce...,NaN,NaN,Experienced (Non-Manager),Bachelor's Degree,Confidential,Full Time | On-site,IT/Software Development | Engineering - Teleco...,ERP | Accounting | Financial Analysis | Oracle...,Job Summary\nWe are looking for a skilled Intu...,Job Requirements\nBachelor’s degree in Account...,2026-04-19T02:03:20
freq,457,9,60,339,127,1,NaN,NaN,325,314,362,314,117,2,2,2,1
mean,NaN,NaN,NaN,NaN,NaN,NaN,3.127193,5.541667,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,2.327392,3.124947,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,0.000000,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,2.000000,3.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,3.000000,5.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,4.000000,7.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# PART 1 - Essential problems in Data

### Problem 1 - Skills Column: Case Duplicates & Fragmented Word Tokens

**Sub-problem A:** The scraper collects skills in the order WUZZUF renders them, which sometimes
includes the same skill twice: once from the "required" chip list and once from the
"preferred" list, with different casing.

Evidence:
```
Row skills: ['Information Technology (IT)', ... , 'Information technology (it)']
            ['Hardware Troubleshooting',    ... , 'Hardware troubleshooting']
```

**Sub-problem B:** The top "skills" by raw frequency are:  `|` (3979), `Development` (258),
`Software` (252), `Computer` (234) — these are NOT skills, they are word fragments
from multi-word skill names that were split incorrectly at some point in the pipeline.
The mean skill count per job is 25.7 with a max of 108, which is unrealistically high.

In [101]:
SKILL_STOPWORDS = {'development', 'software', 'computer', 'science', 'it',
                   'information', 'technology', 'engineering', 'management'}

def filter_fragments(skills_str):
    skills =[s.strip().title() for s in str(skills_str).split(' | ') if s.strip()]
    skills =[s for s in skills if s.lower() not in SKILL_STOPWORDS and len(s) > 2]
    return [' , '.join(dict.fromkeys(skills))]

data['skills'] = data['skills'].apply(filter_fragments)
data['skills']

0      [Customer Service , Customer Support , Custome...
1      [Mechatronics , Accounting , Sales , Mechanica...
2      [Erp , Accounting , Financial Analysis , Oracl...
3                              [Design , Product Design]
4      [Computer Science , Information Technology (It...
                             ...                        
452    [Low Current Engineer , Elv Systems , Fiber Ne...
453    [Uat , Testing , Automation , Wordpress Cms , ...
454    [Professional Content , Reports And Insights ,...
455    [It Infrastructure Management , Artificial Int...
456    [Manual Testing , Test Planning , Performance ...
Name: skills, Length: 457, dtype: object

In [102]:
# Preprocess step #1: Separate Technical Skills from Soft/Generic Skills

# Define the noise terms — generic labels that are categories, not skills
GENERIC_NOISE = {
    'Information Technology (IT)', 'Information Technology',
    'Computer Science', 'Software Development', 'Software Engineering',
    'Software', 'Development', 'Engineering', 'Technology', 'IT',
    'Technical', 'Management', 'Business', 'Science', 'Information',
    'Computer', 'Systems', 'Administration', '(IT)', 'Design',
}

# Define soft skills — valuable for separate soft-skills analysis
SOFT_SKILLS = {
    'Communication', 'Communication Skills', 'Problem Solving', 'Problem-solving',
    'Teamwork', 'Time Management', 'Time management', 'Presentation',
    'Leadership', 'Critical Thinking', 'Attention To Detail', 'Adaptability',
    'Customer Service', 'Customer Support', 'Sales', 'Negotiation',
    'Project Management',
}

def split_skills(skills_str):
    all_skills   = [s.strip() for s in str(skills_str).split(',') if s.strip()]
    technical    = [s for s in all_skills
                    if s not in GENERIC_NOISE and s not in SOFT_SKILLS
                    and len(s) > 2 and not s.isdigit()]
    soft         = [s for s in all_skills if s in SOFT_SKILLS]
    generic      = [s for s in all_skills if s in GENERIC_NOISE]
    return technical, soft, generic

# Apply and create three separate columns
results = data['skills'].apply(split_skills)
data['technical_skills'] = results.apply(lambda x: ', '.join(x[0]))
data['soft_skills']      = results.apply(lambda x: ', '.join(x[1]))
data['skill_count']      = results.apply(lambda x: len(x[0]))  # count of real tech skills

In [103]:
# Preprocess step #2: Skill Alias Normalization

# These are all the same thing — without this, each counts separately
SKILL_ALIASES = {
    # Casing variants
    'Problem-solving':          'Problem Solving',
    'problem solving':          'Problem Solving',
    'Time management':          'Time Management',
    'Customer service':         'Customer Service',
    'Help desk support':        'Help Desk',
    'HelpDesk':                 'Help Desk',
    'Information technology (it)': 'Information Technology (IT)',
    # Tool name variants
    'React.js':                 'React',
    'React js':                 'React',
    'Reactjs':                  'React',
    'Node.js':                  'Node.js',        # keep .js convention
    'NodeJS':                   'Node.js',
    'JavaScript (ES6+)':        'JavaScript',
    'JS':                       'JavaScript',
    'Postgres':                 'PostgreSQL',
    'MS SQL':                   'SQL Server',
    'MSSQL':                    'SQL Server',
    'Cyber Security':           'Cybersecurity',
    'CyberSecurity':            'Cybersecurity',
    'Cyber security':           'Cybersecurity',
    'Quality Assurance':        'QA',
    'Quality Control':          'QC',
    'OOP':                      'Object-Oriented Programming',
    'REST':                     'RESTful APIs',
    'RESTful APIs':             'RESTful APIs',
    'Rest APIs':                'RESTful APIs',
    'ML':                       'Machine Learning',
    'AI':                       'Artificial Intelligence',
}

def normalize_skills(skills_str):
    skills = [s.strip() for s in str(skills_str).split(',') if s.strip()]
    normalized = [SKILL_ALIASES.get(s, s) for s in skills]
    return ', '.join(dict.fromkeys(normalized))  # deduplicate after normalization

data['technical_skills'] = data['technical_skills'].apply(normalize_skills)

In [104]:
# Preprocess step #3: Skill Alias Normalization

data['skill_count'] = data['technical_skills'].apply(
    lambda s: len([x for x in s.split(',') if x.strip()]) if s else 0
)

data[['skills', 'technical_skills', 'soft_skills', 'skill_count']]

,skills,technical_skills,soft_skills,skill_count
0,"[Customer Service , Customer Support , Custome...","['Customer Service, Customer Care, Customer Re...","Customer Support, Sales",32
1,"[Mechatronics , Accounting , Sales , Mechanica...","['Mechatronics, Accounting, Mechanical Enginee...",Sales,7
2,"[Erp , Accounting , Financial Analysis , Oracl...","['Erp, Accounting, Financial Analysis, Oracle,...",,8
3,"[Design , Product Design]","['Design, Product Design']",,2
4,"[Computer Science , Information Technology (It...","['Computer Science, Information Technology (It...",,3
...,...,...,...,...
452,"[Low Current Engineer , Elv Systems , Fiber Ne...","['Low Current Engineer, Elv Systems, Fiber Net...",,5
453,"[Uat , Testing , Automation , Wordpress Cms , ...","['Uat, Testing, Automation, Wordpress Cms, Vis...",,7
454,"[Professional Content , Reports And Insights ,...","['Professional Content, Reports And Insights, ...",,5
455,"[It Infrastructure Management , Artificial Int...","['It Infrastructure Management, Artificial Int...",Project Management,9


### Problem 2 — 51 Genuinely Non-IT Jobs Leaked Into Dataset

This dataset was scraped by browsing the **IT/Software Development** category page this caused
51 roles (11.2% of the dataset) to slip through that are NOT relevant to IT/SW
analysis. Categories include:

| Type | Count | Examples |
|---|---|---|
| Sales & Pre-Sales | 20 | Sales Engineer, IT Sales Manager, Pre-Sales Engineer |
| HR & Recruiting | 3 | HR Manager, Junior Recruiter, Technical Recruiter |
| Non-IT Engineering | 4 | Mechanical Design, Electrical Technical Office |
| Customer Service | 4 | Call Center Agent, Telesales, Technical Support (Call Center) |
| Finance/Admin | 4 | ERP Accountant, Document Controller, Purchasing Specialist |
| Teaching/Training | 3 | ICT Teacher, AI Instructor, Computer Instructor |
| Marketing | 3 | Growth Marketing, Marketing & BD Team Lead |
| Quality Control | 3 | QC Engineer (non-software) |
| Other | 7 | Various non-core roles |

In [105]:
NON_IT_KEYWORDS = [
    'call center', 'customer service', 'telesales', 'outbound reach',
    'sales specialist', 'sales account', 'sales executive', 'sales manager',
    'sales representative', 'sales lead', 'sales software', 'it sales',
    'pre-sales', 'presales', 'hardware sales',
    'hr manager', 'recruiter', 'recruitment coordinator',
    'instructor', 'teacher', 'training',
    'document controller', 'purchasing specialist',
    'marketing specialist', 'marketing team lead',
    'quality control engineer',  # non-software QC
    'mechanical design', 'electrical technical office',
    'erp accountant', 'financial help desk',
    'rta - wfm',
]
def is_non_it(title):
    t = str(title).lower()
    return any(k in t for k in NON_IT_KEYWORDS)

data = data[~data['job_title'].apply(is_non_it)].reset_index(drop=True)

In [106]:
print("Data after Removal of Non-IT jobs: ", data.shape[0])

Data after Removal of Non-IT jobs:  411


### Problem 3 — Salary Column: Mixed Data
The salary column has 4 distinct data types mixed in one field:
1. `'Confidential'` — 362 rows (79.2%)
2. `'Confidential, Benefits...'` — 51 rows (benefits/perks mixed in)
3. `'X to Y CUR Per Period'` — 44 rows (actual numeric salary)
4. Benefits text contains **typos**: `'Bouns'`, `'Boness'` (instead of "Bonus")

**New anomalies vs. previous dataset:**
- **HKD currency** (Hong Kong Dollar): `'15000 to 20000 HKD Per Month'` for a Cairo-based QA Associate at Mico World Limited — likely a foreign company paying in home currency. Needs special handling in normalization.
- **2 Per-Year salaries**: one in EGP (`363300 to 525000`), one in EUR (`19729 to 22054`). Must divide by 12 when normalizing to monthly.
- **1 Per-Hour salary**: `'700 to 1000 EGP Per Hour'` (AI Instructor). Must multiply by ~160 (working hours/month) to normalize.
- **2 flat/pinpoint salaries**: `min == max` (e.g., `10000 to 10000 EGP`) — these are single fixed values, not ranges.
- **Confidential, Confidential, ...** — one row has "Confidential" duplicated inside itself: `'Confidential, Confidential, Hybrid Work Mode...'`

In [107]:
# Normalizing Confidential values to a common one
print("Confidential count before normalization: ", data[data['salary'] == "Confidential"].shape[0])

Confidential count before normalization:  329


In [108]:
data['salary'] = data['salary'].where(
    ~data['salary'].str.startswith('Confidential', na=False),
    other='Confidential'
)

In [110]:
print("Confidential count after normalization: ", data[data['salary'] == "Confidential"].shape[0])

Confidential count after normalization:  375


In [111]:
# Parsing numeric values of salaries
sal_pat = r'([\d,]+)\s+to\s+([\d,]+)\s+(\w+)\s+Per\s+(\w+)'
parsed = data['salary'].str.extract(sal_pat)
parsed.columns = ['sal_min','sal_max','currency','pay_period']
data[['currency','pay_period']] = parsed[['currency','pay_period']]
data['sal_min'] = parsed['sal_min'].str.replace(',','').astype(float)
data['sal_max'] = parsed['sal_max'].str.replace(',','').astype(float)
data['salary_disclosed'] = data['salary'] != 'Confidential'

In [112]:
# Exchange rates as of April 2026 — update as needed
EGP_RATES = {'EGP': 1.0, 'USD': 50.9, 'EUR': 55.2, 'HKD': 6.5, 'AED': 13.9}
PERIOD_MULTIPLIERS = {'Month': 1, 'Year': 1/12, 'Hour': 160}

data['sal_min_egp'] = (
    data['sal_min']
    * data['currency'].map(EGP_RATES).fillna(1)
    * data['pay_period'].map(PERIOD_MULTIPLIERS).fillna(1)
)
data['sal_max_egp'] = (
    data['sal_max']
    * data['currency'].map(EGP_RATES).fillna(1)
    * data['pay_period'].map(PERIOD_MULTIPLIERS).fillna(1)
)

data[data['salary_disclosed'] == True]

,search_query,job_title,company_name,location,posting_date,job_url,years_exp_min,years_exp_max,career_level,education_level,...,technical_skills,soft_skills,skill_count,currency,pay_period,sal_min,sal_max,salary_disclosed,sal_min_egp,sal_max_egp
2,IT-Software-Development,Product Designer,Rescounts,"Cairo, EG",posted 2 days ago,https://wuzzuf.net/jobs/p/yvumfgn0o4yp-product...,5.0,7.0,Experienced (Non-Manager),Not Specified,...,"['Design, Product Design']",,2,EGP,Month,65000.0,75000.0,True,65000.0,75000.0
21,IT-Software-Development,IT Section Head,National Glass,"Sharqia, EG",posted 3 days ago,https://wuzzuf.net/jobs/p/9t4kor20omlf-it-sect...,5.0,10.0,Manager,Bachelor's Degree,...,"['Microsoft Office, Team Leadership, It Infras...","Project Management, Problem Solving, Communica...",8,EGP,Month,20000.0,30000.0,True,20000.0,30000.0
31,IT-Software-Development,Software Engineer (Technical Content),"CC Global Group, LLC","Cairo, EG",posted 5 days ago,https://wuzzuf.net/jobs/p/9rhh6bv7kdib-softwar...,NaN,NaN,Entry Level (Junior Level / Fresh Grad),Bachelor's Degree,...,"['Technical Writing, Content Development, Prog...",,7,EGP,Month,30000.0,40000.0,True,30000.0,40000.0
53,IT-Software-Development,Product Owner / Business Analysis - Fresh Grad...,Alexandria For Programming,"Alexandria, EG",posted 10 days ago,https://wuzzuf.net/jobs/p/utj9rt4djcag-product...,0.0,1.0,Entry Level (Junior Level / Fresh Grad),Not Specified,...,"['Agile, Business Analysis, Analysis, User Sto...",Communication,11,EGP,Month,10000.0,10000.0,True,10000.0,10000.0
58,IT-Software-Development,Splunk Administrator Engineer,Digital cloud for Technology and cybersecurity,"Cairo, EG",posted 11 days ago,https://wuzzuf.net/jobs/p/4foi9uwgik5k-splunk-...,2.0,7.0,Experienced (Non-Manager),Bachelor's Degree,...,"['Soc Analyst L2, Cybersecurity, Siem, Splunk,...",,12,EGP,Month,17000.0,60000.0,True,17000.0,60000.0
102,IT-Software-Development,Software Engineer,Confidential,"Cairo, EG",posted 16 days ago,https://wuzzuf.net/jobs/p/boc8mfuiz0wx-softwar...,0.0,8.0,Entry Level (Junior Level / Fresh Grad),Bachelor's Degree,...,"['Software Engineering, Information Technology...",,2,USD,Month,1000.0,5000.0,True,50900.0,254500.0
105,IT-Software-Development,Quality Assurance Associate,Mico World Limited,"Cairo, EG",posted 17 days ago,https://wuzzuf.net/jobs/p/eiquaedtdrhw-quality...,3.0,3.0,Experienced (Non-Manager),Bachelor's Degree,...,"['Quality Assurance, Quality Control']",,2,HKD,Month,15000.0,20000.0,True,97500.0,130000.0
106,IT-Software-Development,AI Developer (Python) - midlevel,Afaqy,"Cairo, EG",posted 17 days ago,https://wuzzuf.net/jobs/p/xsl2ftrivhne-ai-deve...,3.0,6.0,Experienced (Non-Manager),Not Specified,...,"['Python, Machine Learning, Deep Learning, Llm...",,5,EGP,Month,15000.0,70000.0,True,15000.0,70000.0
111,IT-Software-Development,Junior Full Stack AI Native Developer-Remote,GNC Management Services,"Cairo, EG",posted 17 days ago,https://wuzzuf.net/jobs/p/dydwwxc98muq-junior-...,1.0,2.0,Entry Level (Junior Level / Fresh Grad),Bachelor's Degree,...,"['Python Development, Full Stack Development, ...",,11,USD,Month,500.0,650.0,True,25450.0,33085.0
128,IT-Software-Development,AI Automation & Systems Engineer,Big Life Journal,"Cairo, EG",posted 20 days ago,https://wuzzuf.net/jobs/p/e3sv68tqkom1-ai-auto...,0.0,2.0,Entry Level (Junior Level / Fresh Grad),Bachelor's Degree,...,"['Automation, Html, Information Systems, Infor...",,4,USD,Month,250.0,350.0,True,12725.0,17815.0


### Problem 4 — Posting Date Is Relative Text

The posting_date column contains relative text like `posted 2 hours ago`, `posted 2 days ago`, or `posted 1 month ago`. This is useless for time-series analysis.

In [114]:
import re
from datetime import datetime, timedelta

SCRAPE_DATE = datetime(2026, 4, 19)

def parse_relative_date(s):
    m = re.search(r'(\d+)\s+(hour|day|week|month)', str(s))
    if not m: return pd.NaT
    n, unit = int(m.group(1)), m.group(2)
    deltas = {'hour': timedelta(hours=n), 'day': timedelta(days=n),
              'week': timedelta(weeks=n), 'month': timedelta(days=n*30)}
    return SCRAPE_DATE - deltas[unit]

data['posting_date_parsed'] = data['posting_date'].apply(parse_relative_date)
data['days_since_posted']   = (SCRAPE_DATE - data['posting_date_parsed']).dt.days

data[['posting_date', 'scraped_at', 'posting_date_parsed', 'days_since_posted']]

,posting_date,scraped_at,posting_date_parsed,days_since_posted
0,posted 9 hours ago,2026-04-19T02:03:28,2026-04-18 15:00:00,0
1,posted 11 hours ago,2026-04-19T02:03:35,2026-04-18 13:00:00,0
2,posted 2 days ago,2026-04-19T02:03:41,2026-04-17 00:00:00,2
3,posted 2 days ago,2026-04-19T02:03:48,2026-04-17 00:00:00,2
4,posted 2 days ago,2026-04-19T02:04:01,2026-04-17 00:00:00,2
...,...,...,...,...
406,posted 2 months ago,2026-04-19T02:57:22,2026-02-18 00:00:00,60
407,posted 2 months ago,2026-04-19T02:57:29,2026-02-18 00:00:00,60
408,posted 2 months ago,2026-04-19T02:57:34,2026-02-18 00:00:00,60
409,posted 2 months ago,2026-04-19T02:57:40,2026-02-18 00:00:00,60


### Problem 6 — Location Format Inconsistency

One row uses `'Cairo, Egypt'` (full country name) while all others use the
ISO 2-letter code `'Cairo, EG'`. This causes groupby on location to split
what should be one group into two.

In [115]:
data['city'] = data['location'].str.split(',').str[0].str.strip()

data[['location', 'city']]

,location,city
0,"Damietta, EG",Damietta
1,"Cairo, EG",Cairo
2,"Cairo, EG",Cairo
3,"Giza, EG",Giza
4,"Cairo, EG",Cairo
...,...,...
406,"Alexandria, EG",Alexandria
407,"Cairo, EG",Cairo
408,"Cairo, EG",Cairo
409,"Cairo, EG",Cairo


### Problem 7 — Experience: float64 Dtype + Pinpoint Ranges + High Outliers

**Sub-problem A — float64 dtype:** Same as previous dataset. 1 NaN forces float storage.

**Sub-problem B — 98 rows have min == max (21.4%):**
WUZZUF posts like "Exactly 3 years" which the scraper correctly reads as `(3, 3)`.
These are valid — just flag them so analysis handles them correctly.

**Sub-problem C — 4 high-experience outliers (15–20 years):**
These are legitimate senior roles (CIO, IT Manager). They are valid data points
but will skew average experience calculations significantly. Flag them.

In [116]:
data['years_exp_min'] = data['years_exp_min'].astype('Int64')
data['years_exp_max'] = data['years_exp_max'].astype('Int64')

data[['years_exp_min', 'years_exp_max']]

,years_exp_min,years_exp_max
0,0,2
1,3,5
2,5,7
3,3,7
4,1,3
...,...,...
406,2,2
407,2,3
408,5,5
409,3,10


### Problem 8 — Job Title Quality Issues

**Sub-problem A — 1 fully lowercase title:**
`'odoo senior implementor'` → should be `'Odoo Senior Implementor'`

**Sub-problem B — 6 titles with trailing periods:**
`'Middle Automation QA Engineer.'`, `'Senior Oracle Developer (E-Business Suite).'`, etc.

**Sub-problem C — Inconsistent seniority label conventions in titles:**
`'Middle'` used instead of `'Mid'` (`'Middle Data Engineer'`, `'Middle IOS Developer'`),
`'midlevel'` as one word (`'AI Developer (Python) - midlevel'`).
These all mean the same thing but will create separate groups in aggregations.

In [117]:
data['job_title']

0                           Mechatronics Engineer
1                  Information Systems Specialist
2                                Product Designer
3                            Senior Data Engineer
4                                         Flutter
                          ...                    
406                          Low Current Engineer
407    UAT & Certification Specialist "Outsource"
408         Consultant - Command & Control Center
409                              IT & AI Engineer
410    Senior Software Quality Assurance Engineer
Name: job_title, Length: 411, dtype: object

In [118]:
# Strip trailing punctuation
# Enhance using experience
data['job_title'] = data['job_title'].str.strip().str.rstrip('.!?- ')

# Normalize seniority terms
data['job_title'] = data['job_title'].str.replace(
    r'\bMiddle\b', 'Mid', regex=True
).str.replace(
    r'\bmidlevel\b', 'Mid Level', case=False, regex=True
).str.replace(
    r'\bSr\.\b|\bSr\b', 'Senior', regex=True
).str.replace(
    r'\bJr\.\b|\bJr\b', 'Junior', regex=True
)

data['job_title']

0                           Mechatronics Engineer
1                  Information Systems Specialist
2                                Product Designer
3                            Senior Data Engineer
4                                         Flutter
                          ...                    
406                          Low Current Engineer
407    UAT & Certification Specialist "Outsource"
408         Consultant - Command & Control Center
409                              IT & AI Engineer
410    Senior Software Quality Assurance Engineer
Name: job_title, Length: 411, dtype: object

### Problem 9 - multivalues employment_type
Looking at the `employement_type` column carefully, this column is actually two completely different concepts packed together with `|`:

**Dimension 1 — Contract type:**
Full Time, Part Time, Freelance / Project, Shift Based, Volunteering

**Dimension 2 — Workplace mode:**
 On-site, Hybrid, Remote

In [119]:
data['employment_type']

0      Full Time | Part Time | Volunteering | On-site
1                                 Full Time | On-site
2                                  Full Time | Remote
3                                 Full Time | On-site
4                                  Full Time | Hybrid
                            ...                      
406                               Full Time | On-site
407                               Full Time | On-site
408                               Full Time | On-site
409                               Full Time | On-site
410                                Full Time | Hybrid
Name: employment_type, Length: 411, dtype: object

In [121]:
# ── Step 1: Extract workplace mode (most analytically valuable) ──────
def get_workplace_mode(s):
    s = str(s)
    has_remote  = 'Remote'  in s
    has_hybrid  = 'Hybrid'  in s
    has_onsite  = 'On-site' in s

    if has_remote and not has_hybrid and not has_onsite:
        return 'Fully Remote'
    if has_hybrid:
        return 'Hybrid'
    if has_onsite:
        return 'On-site'
    return 'Unspecified'

data['workplace_mode'] = data['employment_type'].apply(get_workplace_mode)

# ── Step 2: Extract contract type (simplify to dominant) ─────────────
def get_contract_type(s):
    s = str(s)
    # Order matters: Full Time > Part Time > Freelance > others
    if 'Full Time' in s:
        return 'Full Time'
    if 'Part Time' in s:
        return 'Part Time'
    if 'Freelance / Project' in s:
        return 'Freelance'
    if 'Shift Based' in s:
        return 'Shift Based'
    if 'Volunteering' in s:
        return 'Volunteering'
    return 'Unspecified'

data['contract_type'] = data['employment_type'].apply(get_contract_type)

# ── Step 3: Boolean flags for precise multi-label rows ────────────────
# Use these when a single job explicitly advertises multiple options
data['is_full_time']  = data['employment_type'].str.contains('Full Time',        na=False)
data['is_part_time']  = data['employment_type'].str.contains('Part Time',        na=False)
data['is_freelance']  = data['employment_type'].str.contains('Freelance',        na=False)
data['is_remote']     = data['employment_type'].str.contains('Remote',           na=False)
data['is_hybrid']     = data['employment_type'].str.contains('Hybrid',           na=False)
data['is_onsite']     = data['employment_type'].str.contains('On-site',          na=False)

data[['employment_type', 'workplace_mode', 'is_full_time', 'is_part_time', 'is_freelance', 'is_remote', 'is_hybrid', 'is_onsite']]

,employment_type,workplace_mode,is_full_time,is_part_time,is_freelance,is_remote,is_hybrid,is_onsite
0,Full Time | Part Time | Volunteering | On-site,On-site,True,True,False,False,False,True
1,Full Time | On-site,On-site,True,False,False,False,False,True
2,Full Time | Remote,Fully Remote,True,False,False,True,False,False
3,Full Time | On-site,On-site,True,False,False,False,False,True
4,Full Time | Hybrid,Hybrid,True,False,False,False,True,False
...,...,...,...,...,...,...,...,...
406,Full Time | On-site,On-site,True,False,False,False,False,True
407,Full Time | On-site,On-site,True,False,False,False,False,True
408,Full Time | On-site,On-site,True,False,False,False,False,True
409,Full Time | On-site,On-site,True,False,False,False,False,True


### Problem 10 - `job_category` stacked values

Extract Primary + Secondary Signal
From the screenshot, the structure is always:

`[Primary Category] | [Secondary] | [Tertiary...]`

And 70%+ of rows start with IT/Software Development as the primary — meaning the primary category tells you almost nothing unique.

What is valuable is the secondary category, because it signals the domain/industry of the IT role:

- IT/Software Development | Sales/Retail       → IT Sales role
- IT/Software Development | Quality            → QA/Testing role
- IT/Software Development | Project/Program    → PM/Scrum role
- Installation/Maintenance | IT/Software Dev   → Field IT / Support role

In [123]:
# ── Step 1: Normalize separator first ────────────────────────────────
data['job_categories'] = data['job_categories'].str.replace(
    r'\s*\|\s*', ' | ', regex=True
).str.strip()

# ── Step 2: Extract primary category (everything before first |) ──────
data['primary_category'] = data['job_categories'].str.split(' | ').str[0].str.strip()

# ── Step 3: Extract the non-IT secondary category as domain signal ────
IT_CATS = {'IT/Software Development', 'Engineering - Telecom/Technology',
           'Engineering - Other'}

def get_secondary_domain(cat_str):
    cats = [c.strip() for c in str(cat_str).split(' | ')]
    non_it = [c for c in cats if c not in IT_CATS]
    return non_it[0] if non_it else 'Pure IT'

data['job_domain'] = data['job_categories'].apply(get_secondary_domain)

# ── Step 4: Group secondary domains into meaningful buckets ───────────
DOMAIN_MAP = {
    'Pure IT':                      'Core IT',
    'Project/Program Management':   'IT Management',
    'Quality':                      'QA/Testing',
    'Creative/Design/Art':          'Design/UX',
    'Analyst/Research':             'Data/Analytics',
    'Installation/Maintenance/Repair': 'Field/Support IT',
    'Customer Service/Support':     'IT Support',
    'Sales/Retail':                 'IT Sales',
    'Business Development':         'IT Sales',
    'Accounting/Finance':           'IT Finance/ERP',
    'Marketing/PR/Advertising':     'IT Marketing',
    'Logistics/Supply Chain':       'IT Operations',
    'Purchasing/Procurement':       'IT Operations',
    'Operations/Management':        'IT Management',
    'Human Resources':              'HR Tech',
    'Training/Instructor':          'IT Training',
}
data['job_domain_group'] = data['job_domain'].map(DOMAIN_MAP).fillna('Other IT')

data[['job_categories', 'primary_category', 'job_domain', 'job_domain_group']]

,job_categories,primary_category,job_domain,job_domain_group
0,Installation/Maintenance/Repair | IT/Software ...,Installation/Maintenance/Repair,Installation/Maintenance/Repair,Field/Support IT
1,Accounting/Finance | IT/Software Development,Accounting/Finance,Accounting/Finance,IT Finance/ERP
2,Creative/Design/Art | IT/Software Development,Creative/Design/Art,Creative/Design/Art,Design/UX
3,IT/Software Development | Engineering - Teleco...,IT/Software,Pure IT,Core IT
4,Creative/Design/Art | IT/Software Development ...,Creative/Design/Art,Creative/Design/Art,Design/UX
...,...,...,...,...
406,IT/Software Development | Engineering - Teleco...,IT/Software,Pure IT,Core IT
407,Accounting/Finance | IT/Software Development |...,Accounting/Finance,Accounting/Finance,IT Finance/ERP
408,IT/Software Development,IT/Software,Pure IT,Core IT
409,Installation/Maintenance/Repair | IT/Software ...,Installation/Maintenance/Repair,Installation/Maintenance/Repair,Field/Support IT


### Problem 11 — Remove Extra redundant columns after cleaning processes

In [124]:
data.columns

Index(['search_query', 'job_title', 'company_name', 'location', 'posting_date',
       'job_url', 'years_exp_min', 'years_exp_max', 'career_level',
       'education_level', 'salary', 'employment_type', 'job_categories',
       'skills', 'job_description', 'job_requirements', 'scraped_at',
       'technical_skills', 'soft_skills', 'skill_count', 'currency',
       'pay_period', 'sal_min', 'sal_max', 'salary_disclosed', 'sal_min_egp',
       'sal_max_egp', 'posting_date_parsed', 'days_since_posted', 'city',
       'workplace_mode', 'contract_type', 'is_full_time', 'is_part_time',
       'is_freelance', 'is_remote', 'is_hybrid', 'is_onsite',
       'primary_category', 'job_domain', 'job_domain_group'],
      dtype='object')

In [125]:
data = data.drop(columns=['search_query', 'location', 'posting_date', 'job_url', 'salary', 'employment_type', 'job_description', 'job_requirements', 'scraped_at'])

print("Final Data Columns: ", data.columns)

Final Data Columns:  Index(['job_title', 'company_name', 'years_exp_min', 'years_exp_max',
       'career_level', 'education_level', 'job_categories', 'skills',
       'technical_skills', 'soft_skills', 'skill_count', 'currency',
       'pay_period', 'sal_min', 'sal_max', 'salary_disclosed', 'sal_min_egp',
       'sal_max_egp', 'posting_date_parsed', 'days_since_posted', 'city',
       'workplace_mode', 'contract_type', 'is_full_time', 'is_part_time',
       'is_freelance', 'is_remote', 'is_hybrid', 'is_onsite',
       'primary_category', 'job_domain', 'job_domain_group'],
      dtype='object')


# PART 2 - Additional Improvements Specific to Our Dataset

### Addition 1 — education_level: Fill "Not Specified" Intelligently

data has Bachelor's Degree (314 rows) and Not Specified (143 rows = 31.3%).

"Not Specified" isn't truly unknown — it means WUZZUF didn't display a requirement, which itself is a signal (usually means the job is experience-focused, not degree-focused).

In [126]:
data['education_level'] = data['education_level'].replace(
    'Not Specified', 'No Degree Requirement Listed'
)

data['education_level'].unique()

array(['No Degree Requirement Listed', "Bachelor's Degree"], dtype=object)

### Addition 2 — company_name + Confidential salary:

"Confidential" is a real pattern in our and data worth tracking — companies that hide their name also tend to hide salary, which lets you analyze disclosure behavior.

In [128]:
data['fully_anonymous'] = (
    (data['company_name'] == 'Confidential') &
    (~data['salary_disclosed'])   # assuming you created this from salary cleaning
)

data[['fully_anonymous', 'company_name', 'salary_disclosed']]

,fully_anonymous,company_name,salary_disclosed
0,False,Elhelb Smart Solutions,False
1,False,Abou Seif Group,False
2,False,Rescounts,True
3,False,GB Corp,False
4,False,Ibn Sina Pharma,False
...,...,...,...
406,False,Roya Developments,False
407,False,Efada Technology,False
408,True,Confidential,False
409,True,Confidential,False


# PART 3 - Ordering columns Logically + Saving

In [131]:
FINAL_ORDER = [
    # Identity
    'job_title', 'company_name', 'company_disclosed',
    'city', 'fully_anonymous',
    # Time
    'posting_date', 'posting_date_parsed', 'days_since_posted',
    # Classification
    'primary_category', 'job_domain', 'job_domain_group',
    'career_level',
    # Experience
    'years_exp_min', 'years_exp_max',
    # Education
    'education_level',
    # Work style
    'contract_type', 'workplace_mode',
    'is_full_time', 'is_part_time', 'is_freelance',
    'is_remote', 'is_hybrid', 'is_onsite',
    # Salary
    'salary_disclosed', 'sal_min', 'sal_max', 'sal_min_egp', 'sal_max_egp',
    'currency', 'pay_period',
    # Skills
    'technical_skills', 'soft_skills', 'skill_count',
    # Raw archive (keep but at the end)
    'employment_type', 'job_categories', 'salary',
    'job_description', 'job_requirements', 'job_url',
]

data = data[[c for c in FINAL_ORDER if c in data.columns]]
data.to_csv("Data/Clean/Clead_Scrapped_Data.csv", index=False, encoding='utf-8-sig')